src/megatron/energon/wrappers/packing_dataset.py

```diff
@@ -0,0 +1,364 @@
+
+import inspect
+from dataclasses import dataclass
+from typing import (
+    Any,
+    Callable,
+    Dict,
+    Generator,
+    Generic,
+    Iterator,
+    List,
+    Optional,
+    Tuple,
+    TypeVar,
+    Union,
+)
+
+from megatron.energon.errors import SYSTEM_EXCEPTIONS, FatalSampleError
+from megatron.energon.flavors.base_dataset import SavableDataset, set_sample_restore_key
+from megatron.energon.worker import WorkerConfig
+from megatron.energon.wrappers._log_exception import log_exception
+from megatron.energon.wrappers.base import (
+    BaseSingleWrapperDataset,
+    BaseSingleWrapperMergedState,
+    BaseSingleWrapperState,
+    SampleIndex,
+    get_sample_restore_key,
+)
+from megatron.energon.wrappers.buffer import (
+    SampleBufferMergedState,
+    SampleBufferState,
+    SavableSampleBuffer,
+)
+from megatron.energon.wrappers.skip import SkipSample
+
+T_sample = TypeVar("T_sample")
+T_batch_sample = TypeVar("T_batch_sample")
+
+
+@dataclass
+class PackingState(BaseSingleWrapperState):
+    reading_buffer: SampleBufferState
+    pre_packing_buffer: SampleBufferState
+    pre_packing_lengths: List[int]
+    pre_packing_index: int
+    final_packing_index: int
+
+
+@dataclass
+class PackingMergedState(BaseSingleWrapperMergedState):
+    reading_buffer: SampleBufferMergedState
+    pre_packing_buffer: SampleBufferMergedState
+    pre_packing_lengths: List[List[int]]
+    pre_packing_index: List[int]
+    final_packing_index: List[int]
+
+
+class PackingDataset(
+    BaseSingleWrapperDataset[T_sample, T_batch_sample], Generic[T_sample, T_batch_sample]
+):
+    """This dataset wrapper transforms samples of a dataset into chunks/packs of samples, which are
+    then combined into a batch."""
+
+    buffer_size: int
+    pre_packer: Callable[[List[T_sample]], List[List[T_sample]]]
+    final_packer: Callable[[List[T_sample]], T_batch_sample]
+    final_packer_stateless: bool
+    error_handler: Callable[[Exception, List[T_sample]], None]
+    worker_config: WorkerConfig
+
+    #: The buffer for collecting the samples that shall be packed.
+    _reading_buffer: SavableSampleBuffer
+
+    #: Contains the pre-selected samples to be packed.
+    #: The full buffer will be passed to the pre_packer.
+    _pre_packing_buffer: SavableSampleBuffer
+
+    #: Lengths of the selected groups of samples to be packed together.
+    #: The samples are stored sequentially in the pre_packing_buffer because
+    #: SavableSampleBuffer doesn't support nesting. But to keep the groups
+    #: separate, we need to store the lengths of the groups here.
+    _pre_packing_lengths: List[List[int]]
+
+    #: Sample index for the pre_packer
+    _pre_packing_sample_index: SampleIndex
+
+    #: Sample index for the final_packer
+    _final_packing_sample_index: SampleIndex
+
+    def __init__(
+        self,
+        dataset: SavableDataset[T_sample],
+        buffer_size: int,
+        pre_packer: Callable[[List[T_sample]], List[List[T_sample]]],
```

Comment:
I find `pre_packer` and `final_packer` very unclear. Suggesting `grouper` and `merger`, correspondingly in the TE naming this `pack_group` and `pack_merge`.